pkg/netutils/netutils_test.go

```diff
@@ -0,0 +1,1042 @@
+// Copyright (c) 2019 Multus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package netutils
+
+import (
+	"encoding/json"
+	//"fmt"
+	"net"
+	"testing"
+
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types/020"
+	"github.com/containernetworking/cni/pkg/types/current"
+	"github.com/containernetworking/plugins/pkg/ns"
+	"github.com/containernetworking/plugins/pkg/testutils"
+
+	"github.com/vishvananda/netlink"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+func TestNetutils(t *testing.T) {
+    RegisterFailHandler(Fail)
+    RunSpecs(t, "netutils")
+}
+
+// helper function
+func testAddRoute(link netlink.Link, ip net.IP, mask net.IPMask, gw net.IP) error {
+    dst := &net.IPNet{
+        IP:   ip,
+        Mask: mask,
+    }
+    route := netlink.Route{LinkIndex: link.Attrs().Index, Dst: dst, Gw: gw}
+    err := netlink.RouteAdd(&route)
+    return err
+}
+
+func testAddAddr(link netlink.Link, ip net.IP, mask net.IPMask) error {
+    var address = &net.IPNet{IP: ip, Mask: mask}
+    var addr = &netlink.Addr{IPNet: address}
+    err := netlink.AddrAdd(link, addr)
+    return err
+}
+
+func testGetResultFromCache(data []byte) []byte {
+	var cachedInfo map[string]interface{}
+	err := json.Unmarshal(data, &cachedInfo)
+	Expect(err).NotTo(HaveOccurred())
+
+	// try to get result
+	_, ok := cachedInfo["result"]
+	Expect(ok).To(BeTrue())
+
+	resultJSON, ok := cachedInfo["result"].(map[string]interface{})
+	Expect(ok).To(BeTrue())
+
+	resultByte, err := json.Marshal(resultJSON)
+
+	return resultByte
+}
+
+func test020ResultHasIPv4DefaultRoute(data []byte) bool {
+	resultRaw, err := types020.NewResult(data)
+	Expect(err).NotTo(HaveOccurred())
+
+	result, err := types020.GetResult(resultRaw)
+	Expect(err).NotTo(HaveOccurred())
+
+	for _, r := range result.IP4.Routes {
+		if r.Dst.String() == "0.0.0.0/0" {
+			return true
+		}
+	}
+	return false
+}
+
+func test020ResultHasIPv6DefaultRoute(data []byte) bool {
+	resultRaw, err := types020.NewResult(data)
+	Expect(err).NotTo(HaveOccurred())
+
+	result, err := types020.GetResult(resultRaw)
+	Expect(err).NotTo(HaveOccurred())
+
+	for _, r := range result.IP6.Routes {
+		if r.Dst.String() == "::/0" {
+			return true
+		}
+	}
+	return false
+}
```

Comment:
plenty is duplicated here; you could export a 3rd function with most of the code and re-use it from the IP version wrappers:

```golang
func test020ResultHasIPv4DefaultRoute(data []byte) bool {
    v4RouteSelectorFromResult := func (res types020.Result) []cnitypes.Route { 
        return result.IP4.Routes
    }
    return test020ResultHasDefaultRoute(data, v4RouteSelectorFromResult, "0.0.0.0/0")
}

func test020ResultHasIPv6DefaultRoute(data []byte) bool {
    v6RouteSelectorFromResult := func (res types020.Result) []cnitypes.Route {
        return result.IP6.Routes
    }
    return test020ResultHasDefaultRoute(data, v6RouteSelectorFromResult, "::0/0")
}

func test020ResultHasDefaultRoute(data []byte, routeSelector func (result types020.Result) []cnitypes.Route, defaultRouteStr string) bool {
	resultRaw, err := types020.NewResult(data)
	Expect(err).NotTo(HaveOccurred())

	result, err := types020.GetResult(resultRaw)
	Expect(err).NotTo(HaveOccurred())

	for _, r := range routeSelector(result) {
		if r.Dst.String() == defaultRouteStr {
			return true
		}
	}
	return false
}
```

Reply:
Your code above has less duplicated, however, your code size and my code size are pretty similar, so I suppose it is not a big issue.